### PR TITLE
replacer: evaluate format once and improve perf by ~3x

### PR DESCRIPTION
**Pull request by charlievieth**
### 1. Why is this pull request needed and what does it do?

Why: request logging as it currently exists is incredibly slow and wasteful.  This improves the performance of logging by almost 3x and reduces memory usage by ~8x.

Benchmark results:

```
benchmark                                old ns/op     new ns/op     delta
BenchmarkReplacer-12                     644           324           -49.69%
BenchmarkReplacer_CommonLogFormat-12     4228          1471          -65.21%

benchmark                                old allocs     new allocs     delta
BenchmarkReplacer-12                     8              2              -75.00%
BenchmarkReplacer_CommonLogFormat-12     51             17             -66.67%

benchmark                                old bytes     new bytes     delta
BenchmarkReplacer-12                     240           48            -80.00%
BenchmarkReplacer_CommonLogFormat-12     3723          446           -88.02%
```

### 2. Which issues (if any) are related?

N/A

### 3. Which documentation changes (if any) need to be made?

N/A

### 4. Does this introduce a backward incompatible change or deprecation?

N/A

### CODE REVIEW

File: plugin/pkg/replacer/replacer_test.go [hunk 1]
```diff
@@ -32,6 +37,178 @@ func TestReplacer(t *testing.T) {
...
+	}
+	for i, x := range tests {
+		r := parseFormat(x.Format)
+		if !reflect.DeepEqual(r, x.Expected) {
+			t.Errorf("%d: Expected:\n\t%+v\nGot:\n\t%+v", i, x.Expected, r)
+		}
+	}
+}
+
+func TestParseFormat_Nodes(t *testing.T) {
```
Comment (miekg) [inline plugin/pkg/replacer/replacer_test.go]: -> `TestParseFormatNodes` or does the `_` have any significance here?

File: plugin/pkg/replacer/replacer.go [hunk 2]
```diff
@@ -41,168 +56,247 @@ var labels = []string{
...
-	return b.String()
-}
-
-func boolToString(b bool) string {
-	if b {
-		return "true"
+		return append(b, EmptyValue...)
+	default:
+		// TODO (CEV): consider panicking since this should be impossible
+		return append(b, EmptyValue...)
```
Comment (miekg) [inline plugin/pkg/replacer/replacer.go:113]: can we return an error?

File: plugin/pkg/replacer/replacer.go [hunk 3]
```diff
@@ -41,168 +56,247 @@ var labels = []string{
...
+	value string // Literal value, label or metadata label
+	typ   nodeType
+}
+
+// A replacer is an ordered list of all the nodes in a format.
+type replacer []node
+
+func parseFormat(s string) replacer {
+	// Assume there is a literal between each label - its cheaper to over
+	// allocate once than allocate twice.
```
Comment (miekg) [inline plugin/pkg/replacer/replacer.go:183]: ... because `append` will double the capacity? And does it really matter that much here?

File: plugin/pkg/replacer/replacer.go [hunk 4]
```diff
@@ -41,168 +56,247 @@ var labels = []string{
...
+		// We find the right bracket then backtrack to find the left bracket.
+		// This allows us to handle formats like: "{ {foo} }".
+		j := strings.IndexByte(s, '}')
+		if j < 0 {
+			break
+		}
+		i := strings.LastIndexByte(s[:j], '{')
+		if i < 0 {
+			// Handle: "A } {foo}"
+			//            ^j
```
Comment (miekg) [inline plugin/pkg/replacer/replacer.go]: just `^`? Don't understand the "symbol" `^j`

File: plugin/pkg/replacer/replacer.go [hunk 5]
```diff
@@ -41,168 +56,247 @@ var labels = []string{
...
+// bufPool stores pointers to scratch buffers
+var bufPool = sync.Pool{
+	New: func() interface{} {
+		b := make([]byte, 0, 256)
+		return &b
+	},
+}
+
+func (r replacer) Replace(ctx context.Context, state request.Request, rr *dnstest.Recorder) string {
+	p := bufPool.Get().(*[]byte)
```
Comment (miekg) [inline plugin/pkg/replacer/replacer.go]: you have a pointer to a slice, so you can extend the slice or something? This requires some documentation detailing why it's a pointer to a slice. Probably where the syncmap is declared.

**Review by miekg — COMMENTED** [review]
Nice improvements ! Generally looks good, but I do have a bunch of questions.

File: plugin/pkg/replacer/replacer_test.go (hunk 1 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer_test.go]: Nope, will change.

File: plugin/pkg/replacer/replacer.go (hunk 3 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go:183]: Yes and no the extra allocs don't matter so long as parsing doesn't need to be very fast.  That said, I really don't blindly appending to a slice when the required size can be estimated.

Comparison of pre-allocate (current method) to appending.
```
$ benchcmp pre-allocate.txt append.txt
benchmark                   old ns/op     new ns/op     delta
BenchmarkParseFormat-12     527           918           +74.19%

benchmark                   old allocs     new allocs     delta
BenchmarkParseFormat-12     1              6              +500.00%

benchmark                   old bytes     new bytes     delta
BenchmarkParseFormat-12     704           1520          +115.91%
```

File: plugin/pkg/replacer/replacer.go (hunk 4 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go]: Will fix.

File: plugin/pkg/replacer/replacer.go (hunk 5 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go]: I'll add a comment.  The reason we use store a pointer to the slice and not the slice itself is that it saves an alloc in the value => `interface{}` conversion.

File: plugin/pkg/replacer/replacer.go (hunk 2 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go:113]: This code will only be executed if there is a bug in the `parseFormat()` or `replacer.Replace()` logic and matches the previous behavior of the now removed `value()` function.  

Basically, this would be a hard programming error so I'm not sure there would be much value in propagating this to the user as there is nothing they can do about it (nor is there an easy way to propagate this error).

Personally, I would panic as I like to treat programming/logic errors harshly so that they do not make into production (I consider the behavior of unlocking an unlocked `sync.Mutex` to be a good example of this).  Additionally, if there is a bug here it is impossible to know if the logs being generated are correct.

That said, I'm open to suggestions as to how you think this should be handled.

**TLDR:** I don't think there is much value in returning an error - since this represents a logic bug.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] replacer: eva..." ]
>I'll add a comment.  The reason we use store a pointer to the slice and not the slice itself is that it saves an alloc in the value => `interface{}` conversion.

this sounds like a really micro optimizations we shouldn't really care about. I 
doubt this makes a dent in the 3x improvement we're seeing here.

Logging will always kill your performance, doing these micro optimizations won't 
make it that much better. Just use the slice directly.

/Miek

--
Miek Gieben

File: plugin/pkg/replacer/replacer.go (hunk 5 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go]: Comment added to bufPool https://github.com/coredns/coredns/blob/57fa92527aa00dc0b38db10eda5fd204218f1d46/plugin/pkg/replacer/replacer.go#L270-L274

Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go]: > [ Quoting <notifications@github.com> in "Re: [coredns/coredns] replacer: eva..." ]
> I'll add a comment. The reason we use store a pointer to the slice and not the slice itself is that it saves an alloc in the value => `interface{}` conversion.
> this sounds like a really micro optimizations we shouldn't really care about. I doubt this makes a dent in the 3x improvement we're seeing here. Logging will always kill your performance, doing these micro optimizations won't make it that much better. Just use the slice directly. /Miek
> […](#)
> -- Miek Gieben

IMO, more a matter of correctness, but I'll change it to use the value so we don't have to deal with pointers.

**Comment by miekg** [comment]
>**TLDR:** I don't think there is much value in returning an error - since this represents a logic bug.

ok, then remove the comment, no-one will read logs or anything, so if we can't 
make this a failure on startup it doesn't matter.

**Comment by charlievieth** [comment]
> **TLDR:** I don't think there is much value in returning an error - since this represents a logic bug.
> ok, then remove the comment, no-one will read logs or anything, so if we can't make this a failure on startup it doesn't matter.

Comment removed.  Also, thank you for taking the time to review this and for the quick feedback!

File: plugin/pkg/replacer/replacer.go (hunk 2 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go:113]: Comment removed.

**Comment by charlievieth** [comment]
Thanks again for the quick feedback, please let me know if there are any other changes you'd like for me to make.

File: plugin/pkg/replacer/replacer.go (hunk 2 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go:113]: Note the comment has been removed, but is not showing up in the diff.

File: plugin/pkg/replacer/replacer.go [hunk 6]
```diff
@@ -14,13 +15,27 @@ import (
...
+
+const (
+	headerReplacer = "{>"
+	// EmptyValue is the default empty value.
+	EmptyValue = "-"
+)
+
 // labels are all supported labels that can be used in the default Replacer.
-var labels = []string{
+var labels = [...]string{
```
Comment (miekg) [inline plugin/pkg/replacer/replacer.go]: why the `...` here?

**Review by miekg — CHANGES_REQUESTED** [review]
looks good, one nit.

File: plugin/pkg/replacer/replacer.go (hunk 6 above)
Comment (charlievieth) [reply plugin/pkg/replacer/replacer.go]: That was a nit on my part.  I like to declare what are essentially `const slices` as arrays to indicate to other programers that they are const and not to be modified.

That said, the `labels` var is no longer necessary and I will replace it with the `knownLabels` map.
